exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +317,272 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+
+			patchHelper, err := patch.NewHelper(infraMachine, r.Client)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create patch helper for %s", klog.KObj(infraMachine))
+			}
+
+			ownerRefs = util.EnsureOwnerRef(ownerRefs, *machineRef)
+			infraMachine.SetOwnerReferences(ownerRefs)
+
+			if err := patchHelper.Patch(ctx, infraMachine); err != nil {
+				return errors.Wrapf(err, "failed to patch %s", klog.KObj(infraMachine))
+			}
+		}
+	}
+
+	return nil
+}
+
+// getNewMachine creates a new Machine object. The name of the newly created resource is going
+// to be created by the API server, we set the generateName field.
+func getNewMachine(mp *expv1.MachinePool, infraMachine *unstructured.Unstructured) *clusterv1.Machine {
+	infraRef := corev1.ObjectReference{}
+	if infraMachine != nil {
+		infraRef.APIVersion = infraMachine.GetAPIVersion()
+		infraRef.Kind = infraMachine.GetKind()
+		infraRef.Name = infraMachine.GetName()
+		infraRef.Namespace = infraMachine.GetNamespace()
+	}
+
+	machine := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", mp.Name)),
+			// Note: by setting the ownerRef on creation we signal to the Machine controller that this is not a stand-alone Machine.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(mp, mp.GroupVersionKind())},
+			Namespace:       mp.Namespace,
+			Labels:          make(map[string]string),
+			Annotations:     make(map[string]string),
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName:       mp.Spec.ClusterName,
+			InfrastructureRef: infraRef,
+		},
+	}
+
+	for k, v := range mp.Spec.Template.Annotations {
+		machine.Annotations[k] = v
+	}
+
+	// Set the labels from machinePool.Spec.Template.Labels as labels for the new Machine.
+	// Note: We can't just set `machinePool.Spec.Template.Labels` directly and thus "share" the labels
+	// map between Machine and machinePool.Spec.Template.Labels. This would mean that adding the
+	// MachinePoolNameLabel later on the Machine would also add the labels to machinePool.Spec.Template.Labels
+	// and thus modify the labels of the MachinePool.
+	for k, v := range mp.Spec.Template.Labels {
+		machine.Labels[k] = v
+	}
+
+	// Enforce that the MachinePoolNameLabel and ClusterNameLabel are present on the Machine.
+	machine.Labels[clusterv1.MachinePoolNameLabel] = capilabels.MustFormatValue(mp.Name)
+	machine.Labels[clusterv1.ClusterNameLabel] = mp.Spec.ClusterName
+
+	return machine
+}
+
+// infraMachineToMachinePoolMapper is a mapper function that maps an InfraMachine to the MachinePool that owns it.
+// This is used to trigger an update of the MachinePool when a InfraMachine is changed.
+func infraMachineToMachinePoolMapper(_ context.Context, o client.Object) []ctrl.Request {
+	labels := o.GetLabels()
+
+	if poolName, ok := labels[clusterv1.MachinePoolNameLabel]; ok {
```

Comment:
Is it the same case with annotations? Since we expect to trigger this mapper quite a bit, I think we don't want to make it too expensive with lookups/client calls.

Reply:
Also, I wonder if we would be able to convert the pool name to a label as well and check for a match (so if it's too long they'll both be the same hash).

Reply:
> Is it the same case with annotations?

I don't know.

> Since we expect to trigger this mapper quite a bit, I think we don't want to make it too expensive with lookups/client calls.

Get'ing Machines and MachineSets will all be cached calls (aka they are just read from the local cache). So this will take ~ microseconds.

> Also, I wonder if we would be able to convert the pool name to a label as well and check for a match (so if it's too long they'll both be the same hash).

This should work. Basically you can list all MachinePools (which is also a cached call), then convert the pool names and compare with the InfraMachine. Should be fine.

Please note we still have to change the label selector in l. 344 and document that providers also have to use the `capilabels.MustFormatValue` func when they create InfraMachines.

Reply:
Sounds good, I'll take a shot at trying to match the label/hash. I also prefer this to trying to traverse ownerRefs as that seems a lot more prone to error.

Reply:
Looks mostly good, commented on the implementation: https://github.com/kubernetes-sigs/cluster-api/pull/8828#discussion_r1254555204